# Transposable elements mediate genetic effects altering the expression of nearby genes in colorectal cancer


https://doi.org/10.1101/2021.12.03.471093

## Directory structure 

.
├── Figures -> *This directory contains all figures and the scripts for generating them.*
│   ├── Figure1
│   ├── Figure2
│   ├── Figure3
│   ├── Figure4
│   ├── Figure5
│   ├── Figure6
│   ├── SuppFigure1
│   ├── SuppFigure10
│   ├── SuppFigure11
│   ├── SuppFigure12
│   ├── SuppFigure13
│   ├── SuppFigure14
│   ├── SuppFigure15
│   ├── SuppFigure16
│   ├── SuppFigure17
│   ├── SuppFigure18
│   ├── SuppFigure19
│   ├── SuppFigure2
│   ├── SuppFigure20
│   ├── SuppFigure21
│   ├── SuppFigure3
│   ├── SuppFigure4
│   ├── SuppFigure5
│   ├── SuppFigure6
│   ├── SuppFigure7
│   ├── SuppFigure8
│   └── SuppFigure9
├── analysis_DGE *This directory contains the script used to perform differential expression analysis of our genes/TEs.*
├── analysis_qtlReplication *This directory contains all scripts used to replicate our findings in external datasets.*
├── binaries *Contains the executables for the different software used* 
├── data_covariates *This directory contains the scripts used to generate the covariates used for residualizing our molecular phenotypes.*
├── data_expression *Contains script for bed file normalization adding back the y-intercept to the residuals.* 
├── data_qtls * This directory contains all scripts used to perform all QTL analyses.* 
│   ├── emods
│   │   ├── mapping
│   │   ├── nominal
│   │   ├── permute
│   │   └── quantifications
│   ├── eqtl
│   │   ├── interaction
│   │   ├── mapping
│   │   ├── methylation
│   │   ├── nominal
│   │   ├── optimize
│   │   ├── replicate
│   │   └── specific
│   └── teqtls
│       ├── bn
│       │   └── __pycache__
│       ├── permute
│       └── triplets
└── revisions *This directory contains all scripts used for running the various analyses asked by the reviewers.* 
    ├── analysis_MSS_MSI_TE_expression
    ├── analysis_NeqtlsTEage
    ├── analysis_SNPinTEs
    ├── analysis_TEgenomeJunction
    └── analysis_multimapped

**analysis_DGE**: This directory contains the script used to perform differential expression analysis of our genes/TEs. 
**analysis_qtlReplication**: This directory contains all scripts used to replicate our findings in external datasets. 
**data_covariates**: This directory contains the scripts used to generate the covariates used for residualizing our molecular phenotypes. 
**data_expression**: Contains script for bed file normalization adding back the y-intercept to the residuals. 
**data_qtls**: This directory contains all scripts used to perform all QTL analyses. 
**Figures**: This directory contains all figures and the scripts for generating them. 
**revisions**: This directory contains all scripts used for running the various analyses asked by the reviewers. 



